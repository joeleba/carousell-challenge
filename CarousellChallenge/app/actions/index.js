export const ADD_POST = "ADD_POST";
export const ADD_COMMENT = "ADD_COMMENT";
export const UPVOTE = "UPVOTE";
export const DOWNVOTE = "DOWNVOTE";

let nextContentId = 0
export const addPost = (title, text) => ({
  type: ADD_POST,
  id: nextContentId++,
  title: title,
  text
})

export const addComment = (parent_id, text) => ({
  type: ADD_COMMENT,
  id: nextContentId++,
  parent_id: parent_id,
  text
})

export const upvote = (id) => ({
  type: UPVOTE,
  id
})

export const downvote = (id) => ({
  type: DOWNVOTE,
  id
})
