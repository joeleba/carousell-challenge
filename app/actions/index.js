export const ADD_POST = 'ADD_POST';
export const ADD_COMMENT = 'ADD_COMMENT';
export const UPVOTE = 'UPVOTE';
export const DOWNVOTE = 'DOWNVOTE';
export const POST_TYPE = 'POST';
export const COMMENT_TYPE = 'COMMENT'

export const addPost = (title, text) => ({
  type: ADD_POST,
  title,
  text,
});

export const addComment = (parentId, parentType, text) => ({
  type: ADD_COMMENT,
  parentId,
  parentType,
  text,
});

export const upvote = (id, contentType) => ({
  type: UPVOTE,
  contentType,
  id,
});

export const downvote = (id, contentType) => ({
  type: DOWNVOTE,
  contentType,
  id,
});
