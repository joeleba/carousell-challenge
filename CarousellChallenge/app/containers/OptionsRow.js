import React, { Component } from 'react';
import { AppRegistry, Text, FlatList, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import Icon from 'react-native-vector-icons/FontAwesome';

import { upvote, downvote } from '../actions';
import NewContentForm from '../components/NewContentForm';
import CommentThreadWrapper from '../containers/CommentThread';
import _ from 'lodash'

class OptionsRow extends Component {
  _keyExtractor(item, index) {
    return item.id;
  }

  render() {
    let currentComment = this.props.currentComment;
    let postId = this.props.postId;

    return (
      <View style={styles.optionsRow}>
        <Text
          style={styles.option}
          onPress={() => this.props.dispatchNavigateToNewReply(postId, currentComment)}>
          <Icon name="reply">{" Reply  |  "}</Icon>
        </Text>

        <Text style={styles.option}>
          <Icon
            name="arrow-up"
            onPress={() => this.props.dispatchUpvoteAction(currentComment.id)}></Icon>
          {` ${(currentComment.upvoteCount - currentComment.downvoteCount)} `}
          <Icon
            name="arrow-down"
            onPress={() => this.props.dispatchDownvoteAction(currentComment.id)}></Icon>
        </Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  optionsRow: {
    height: 15,
    alignSelf: 'flex-end',
    flexDirection: 'row',
  },
  option: {
    color: 'grey',
    fontSize: 12
  }
})

function mapStateToProps(state) {
  return {}
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToNewReply: (postId, replyingToObject) =>
      dispatch(NavigationActions.navigate({
        routeName: 'NewComment',
        params: {
          postId: postId,
          replyingToObject: replyingToObject
        }
      })),
    dispatchUpvoteAction: (id) => dispatch(upvote(id)),
    dispatchDownvoteAction: (id) => dispatch(downvote(id))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(OptionsRow)
