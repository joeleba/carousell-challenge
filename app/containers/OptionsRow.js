import React, { Component } from 'react';
import { Text, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import Icon from '@expo/vector-icons/FontAwesome';

import { upvote, downvote } from '../actions';

class OptionsRow extends Component {
  keyExtractor(item, index) {
    return item.id;
  }

  render() {
    let currentComment = this.props.currentComment;
    let postId = this.props.postId;
    let contentType = this.props.contentType;

    return (
      <View style={styles.optionsRow}>
        <Text
          style={styles.option}
          onPress={() => this.props.dispatchNavigateToNewReply(postId, currentComment, contentType)}
        >
          <Icon name="reply">
            {' Reply  |  '}
          </Icon>
        </Text>

        <Text style={styles.option}>
          <Icon
            name="arrow-up"
            onPress={() => this.props.dispatchUpvoteAction(currentComment.id, contentType)}
          />
          {` ${(currentComment.upvoteCount - currentComment.downvoteCount)} `}
          <Icon
            name="arrow-down"
            onPress={() => this.props.dispatchDownvoteAction(currentComment.id, contentType)}
          />
        </Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  optionsRow: {
    height: 15,
    alignSelf: 'flex-end',
    flexDirection: 'row',
  },
  option: {
    color: 'grey',
    fontSize: 12,
  },
});

function mapStateToProps(state) {
  return {};
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToNewReply: (postId, replyingToObject, contentType) =>
      dispatch(NavigationActions.navigate({
        routeName: 'NewComment',
        params: {
          postId,
          replyingToObject,
          contentType
        },
      })),
    dispatchUpvoteAction: (id, contentType) => dispatch(upvote(id, contentType)),
    dispatchDownvoteAction: (id, contentType) => dispatch(downvote(id, contentType)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(OptionsRow);
