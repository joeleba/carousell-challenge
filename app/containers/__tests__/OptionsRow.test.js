import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import OptionsRow from '../OptionsRow';
import { upvote, downvote } from '../../actions';

const mockStore = configureStore();

const stateNoComment = {
  contentReducers: {
    currentId: 1,
    posts: {
      0: {
        children: [],
        content: "ccc",
        downvoteCount: 0,
        upvoteCount: 0,
        title: "sad",
        id: 0
      }
    },
    comments: {},
  },
};

const currentComment = {
  children: [],
  content: "ccc",
  downvoteCount: 0,
  upvoteCount: 0,
  title: "sad",
  id: 0
}

describe('Component: OptionsRow', () => {
  it('renders correctly', () => {
    const store = mockStore(stateNoComment);
    const wrapper = shallow(
      <OptionsRow currentComment={currentComment} postId={0} />,
      { context: { store: store } },
    );

    const render = wrapper.dive();
    expect(render).toMatchSnapshot();
  });

  it('dispatches correct actions on upvote/downvote', () => {
    const store = mockStore(stateNoComment);
    const wrapper = shallow(
      <OptionsRow currentComment={currentComment} postId={0} />,
      { context: { store: store } },
    );
    const render = wrapper.dive();

    render.find('[name="arrow-up"]').first().simulate('press');
    render.find('[name="arrow-down"]').first().simulate('press');

    const actions = store.getActions();
    const expectedPayload = [{"id": 0, "type": "UPVOTE"}, {"id": 0, "type": "DOWNVOTE"}];
    expect(actions).toEqual(expectedPayload);
  });
});
