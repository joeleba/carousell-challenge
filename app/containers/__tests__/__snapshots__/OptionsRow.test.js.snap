// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: OptionsRow renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "flex-end",
      "flexDirection": "row",
      "height": 15,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    onPress={[Function]}
    style={
      Object {
        "color": "grey",
        "fontSize": 12,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      name="reply"
      size={12}
    >
       Reply  |  
    </Icon>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "grey",
        "fontSize": 12,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      name="arrow-up"
      onPress={[Function]}
      size={12}
    />
     0 
    <Icon
      allowFontScaling={false}
      name="arrow-down"
      onPress={[Function]}
      size={12}
    />
  </Text>
</View>
`;
